File: python/root/Visitor.py
from __future__ import annotations
from typing import TYPE_CHECKING
import random

from python.root.Settings import Settings
from python.root.VisitorRequestStatus import VisitorRequestStatus

if TYPE_CHECKING:
    from python.root.Product import Product
    from python.root.CoffeeHouse import CoffeeHouse
    from python.root.CashDesk import CashDesk


class Visitor:
    _model: CoffeeHouse
    _cashDesk: CashDesk
    _desiredProduct: Product
    _nextActionTime: int
    _successRating: int

    def __init__(self, model: CoffeeHouse) -> None:
        self._nextActionTime = 0
        self._successRating = 5
        self._model = model
        menu = model.getMenu()
        self._desiredProduct = menu[random.choice(range(len(menu)))]
        cashDeskList = model.getCashDeskList()
        self._cashDesk = random.choice(cashDeskList)
        self._cashDesk.addVisitorToQueue(self)

    def process(self, currentTime: int) -> None:
        if self._nextActionTime == 0:
            self._nextActionTime = currentTime + Settings.VISITOR_WAITING_TIME
        if currentTime >= self._nextActionTime:
            self._model.addRating(0)
            self._model.visitorLeave(self)
            print(f'Time={currentTime} Visitor leave by timeout {self}')

    def askDesiredProduct(self, currentTime: int) -> None:
        requestStatus = self._model.requestProduct(self._desiredProduct)
        if requestStatus == VisitorRequestStatus.SUCCESS:
            self._model.takeProduct(self._desiredProduct)
            self._model.addRating(self._successRating)
            self._model.visitorLeave(self)
            print(f'Time={currentTime} Visitor({self.__hash__()}) leave with product {self._desiredProduct}')
        elif requestStatus == VisitorRequestStatus.PRODUCT_MISSING:
            self.failDesiredProduct(currentTime, self._desiredProduct)

    def givePreparedProduct(self, currentTime: int, product: Product) -> None:
        self._model.addRating(self._successRating)
        self._model.visitorLeave(self)
        print(f'Time= {currentTime} Visitor({self.__hash__()}) leave with product {product}')

    def failDesiredProduct(self, currentTime: int, product: Product) -> None:
        self._model.addRating(1)
        self._model.visitorLeave(self)
        print(f'Time= {currentTime} Visitor({self.__hash__()}) leave without product {product}')

    def getDesiredProduct(self) -> Product:
        return self._desiredProduct

    def getCashDesk(self) -> CashDesk:
        return self._cashDesk

    def __str__(self) -> str:
        if self in self._cashDesk.getVisitorList():
            return f'Visitor({self.__hash__()})\
             desiredProduct=[{self._desiredProduct}]\
              queuePos={self._cashDesk.getVisitorList().index(self)}'
        else:
            return f'Visitor({self.__hash__()})\
                         desiredProduct=[{self._desiredProduct}]\
                          queuePos: this visitor changed his mood and left the queue'

File: cplusplus/Visitor.cpp
#include "Visitor.hpp"

Visitor::Visitor(CoffeeHouse *model)
{
	this->model = model;

	auto menu = model->getMenu();

	desiredProduct = menu[rand() % menu.size()];

	auto cashDeskVector = model->getCashDeskVector();
	cashDesk = cashDeskVector[rand() % cashDeskVector.size()];
	cashDesk->addVisitorToQueue(this);
}

void Visitor::process(long currentTime)
{
	if (nextActionTime == 0)
	{
		nextActionTime = currentTime + Settings::VISITOR_WAITING_TIME;
	}
	if (currentTime >= nextActionTime)
	{
		model->addRating(0);
		model->visitorLeave(this);
		std::cout << "Time=" + std::to_string(currentTime) + " Visitor leave by timeout " << this << std::endl;
	}
}

void Visitor::askDesiredProduct(long currentTime)
{
	auto requestStatus = model->requestProduct(desiredProduct);
	if (requestStatus == VisitorRequestStatus::SUCCESS)
	{
		model->takeProduct(desiredProduct);
		model->addRating(successRating);
		model->visitorLeave(this);
		std::cout << "Time=" + std::to_string(currentTime) + " Visitor(" + getHashCode() + ") leave with product " << desiredProduct->toString() << std::endl;
	}
	else if (requestStatus == VisitorRequestStatus::PRODUCT_MISSING)
	{
		failDesiredProduct(currentTime, desiredProduct);
	}
}

void Visitor::givePreparedProduct(long currentTime, Product *product)
{
	model->addRating(successRating);
	model->visitorLeave(this);
	std::cout << "Time=" + std::to_string(currentTime) + " Visitor(" + getHashCode() + ") leave with product " << product->toString() << std::endl;
}

void Visitor::failDesiredProduct(long currentTime, Product *product)
{
	model->addRating(1);
	model->visitorLeave(this);
	std::cout << "Time=" + std::to_string(currentTime) + " Visitor(" + getHashCode() + ") leave without product " << product->toString() << std::endl;
}

Product *Visitor::getDesiredProduct()
{
	return desiredProduct;
}

CashDesk *Visitor::getCashDesk()
{
	return cashDesk;
}

std::string Visitor::toString()
{
	std::vector<Visitor *> visitorVector = cashDesk->getVisitorVector();
	auto it = std::find(visitorVector.begin(), visitorVector.end(), this);
	return "Visitor(" + getHashCode() + ") desiredProduct=[" + desiredProduct->toString() + "] queuePos=" + std::to_string(it - visitorVector.begin());
}

std::string Visitor::getHashCode()
{
	std::string thisAddress = std::to_string(reinterpret_cast<int>(this));
	std::string hashCode = std::to_string(std::hash<std::string>()(thisAddress));

	return hashCode;
}
